wire-ios-sync-engine/Source/Calling/WireCallCenterV3.swift

```diff
@@ -664,6 +664,40 @@ extension WireCallCenterV3 {
                 callSnapshots[conversationId] = previousSnapshot.update(with: .terminating(reason: reason))
             }
         }
+
+        leaveSubconversationIfNeeded(id: conversationId)
+    }
+
+    private func leaveSubconversationIfNeeded(id: AVSIdentifier) {
+        guard
+            let context = uiMOC,
+            let syncContext = context.zm_sync,
+            let domain = id.domain ?? BackendInfo.domain,
+            let conversation = ZMConversation.fetch(
+                with: id.identifier,
+                domain: domain,
+                in: context
+            ),
+            conversation.messageProtocol == .mls,
+            let parentGroupID = conversation.mlsGroupID
+        else {
+            return
+        }
+
+        syncContext.perform {
+            guard let mlsService = syncContext.mlsService else {
+                WireLogger.calling.error("failed to leave subconversation: mlsService is missing")
+                return
+            }
+
+            Task {
+                try await mlsService.leaveSubconversation(
+                    parentQualifiedID: QualifiedID(uuid: id.identifier, domain: domain),
+                    parentGroupID: parentGroupID,
+                    subconversationType: .conference
+                )
+            }
+        }
```

Comment:
This implementation seems very similar to `leaveStaleConferenceIfNeeded` in `WireCallCenterV3+MLS`. I see that the main difference is that we don't call the same method from `mlsService` (`leaveSubconversation` vs `leaveSubconversationIfNeeded`). Could you help me understand the difference between the two and why one is used over the other?

Reply:
When we close a call we know we should leave the sub conversation, so we invoke `mlsService.leaveSubconversation` directly. This will leave the subconversation on the backend and then wipe the subgroup in CC. This is the happy path, where the self client is present when the call ends (by either side).

Suppose the self client crashes while in a call. When it relaunches, the call may be ongoing still, but the self client isn't part of the call anymore. It is however still part of the subconversation (backend) and subgroup (CC). To correct this state, we need to detect whether the subconversation has gone stale. This is done for instance by observing an incoming call state, which the self client will receive from the ongoing call. In this case we can detect that we're part of the subconversation but not the call, so we correct the state by leaving the subconversation.

`mlsService.leaveSubconversationIfNeeded` contains the logic for checking if we're part of a subconversation, and assuming that we shouldn't be, will leave the subconversation if we are. It does so by invoking `mlsService.leaveSubconversation`.

Perhaps the naming between the mls service methods and the call center methods is a bit confusing.

Reply:
Yes perhaps renaming the methods or adding comments could help bring more clarity

Reply:
I refactored the methods a bit and changed the names, also added comments.